### ISSUE TITLE
Distributions log_prob method returns None for gradients with respect to some trainable variables, although normal for other variables 

### ISSUE THREAD

**Issue by xiaolong1979** [tensorflow/probability#1351]
I am trying to modify the approach in https://www.tensorflow.org/probability/examples/Linear_Mixed_Effects_Models for a different model but could not get the right gradients to continue. Is there any way to see why the log_prob method returns None for some trainable variables? Here are my colab codes:

import pandas as pd
import numpy as np
from pprint import pprint
import matplotlib.pyplot as plt
import numpy as np
import seaborn as sns

import tensorflow as tf
import tensorflow_probability as tfp
from tensorflow_probability import distributions as tfd
tfd = tfp.distributions
tfb = tfp.bijectors

### generate some sample feature data to feed the model
X=tf.random.uniform(shape=[10,4])
source=tf.one_hot(tf.random.categorical(tf.math.log([[0.5, 0.3, 0.2]]), 10),3)[0,:]
treatment_group= tf.one_hot(tf.random.categorical(tf.math.log([[0.6, 0.4]]), 10),2)[0,:]

### Derive model parameters
num_subjects=X.shape[0]
num_sources=source.shape[1]
num_treatments=treatment_group.shape[1]
num_baseline_features=X.shape[1]

num_targets=4   ### outcome dimension

### build the model
class BayesianWLS(tf.Module):
    def __init__(self, num_targets,num_subjects,num_sources):
      ### trainable parameters
      

      self.std_error=tfp.util.TransformedVariable( 0.01+tf.zeros([num_treatments+num_sources], tf.float32),  
                                                  bijector=tfb.Exp(), name="std_error")
      self.txbeta=tf.Variable(1.01+tf.zeros([num_treatments+num_baseline_features]),name='txbeta')
      self.mu_nu_source=tf.Variable(tf.zeros([num_sources]), name="mean_nu_source")
      self.std_nu=tfp.util.TransformedVariable(0.01+ tf.zeros([1], tf.float32),  
                                               bijector=tfb.Exp(), name="std_nu")
      self.std_eta=tfp.util.TransformedVariable(0.01+ tf.zeros([1], tf.float32),  bijector=tfb.Exp(), name="std_eta")
      self.std_delta=tfp.util.TransformedVariable( 0.01+tf.zeros([1], tf.float32),  bijector=tfb.Exp(), name="std_delta")
      self.beta=tf.Variable(tf.zeros([num_targets]), name="slope_beta")
 
    def __call__(self, X,source,treatment_group):   
        ### get treatment and source specific error stds
        tx_source=tf.concat([tf.cast(treatment_group, tf.float32), source],axis=1) 
        scale_diag=tf.repeat(tf.expand_dims(tf.exp(tf.tensordot( tx_source, self.std_error, axes=((1),(0)))), -1) , num_targets, axis=1)
        
        ### get feature and treatment fixed effect 
        Bcov=tf.concat([tf.cast(treatment_group, tf.float32), X],axis=1)
        alpha_X=tf.repeat(tf.expand_dims(tf.math.exp(
            tf.tensordot( Bcov, self.txbeta, axes=((1),(0)))),-1) , num_targets, axis=1) ### exp(theta+alpha*X)

        ### outcome following the model
        def model():
          nu = yield tfd.MultivariateNormalDiag(loc=tf.repeat(tf.expand_dims(tf.tensordot(source, self.mu_nu_source, axes=((1),(0))), -1), repeats=num_targets, axis=-1), scale_identity_multiplier=1.0)
          eta = yield tfd.Gamma(concentration=1/(self.std_eta)**2+tf.zeros(num_subjects), rate=1/(self.std_eta)**2+tf.zeros(num_subjects))
          delta = yield tfd.Gamma(concentration=1/(self.std_delta)**2+tf.zeros(num_sources), rate=1/(self.std_delta)**2+tf.zeros(num_sources))
          Y = yield tfd.MultivariateNormalDiag(
                    loc=nu+alpha_X*tf.repeat(tf.expand_dims(eta, -1), repeats=num_targets, axis=-1)*tf.repeat(tf.expand_dims(tf.tensordot( source, delta, axes=((1),(0))), -1), repeats=num_targets, axis=-1)*tf.math.cumsum(self.beta)
                  , 
                    scale_diag=scale_diag)
        joint = tfd.JointDistributionCoroutineAutoBatched(model)
        joint._to_track = self
        return joint

WLSclass=BayesianWLS(num_targets,num_subjects,num_sources)
WLSmodel=WLSclass(X,source,treatment_group)

trainable_variables = WLSmodel.trainable_variables
current_state = WLSmodel.sample()[:-1]

### get simulated outcome
ymodel=WLSmodel.sample()[-1]

### as a function of parameter state given the observed target

target_log_prob_fn = lambda *x: WLSmodel.log_prob(x + (ymodel,))

with tf.GradientTape() as tape:
  loss = -target_log_prob_fn(*current_state)
  grads = tape.gradient(loss, trainable_variables)
for i in range(len(grads)):
 print( trainable_variables[i].name, grads[i])

Output problem:

slope_beta:0 tf.Tensor([-446.61765  -189.513     -42.20112    47.819344], shape=(4,), dtype=float32)
mean_nu_source:0 tf.Tensor([ 3.877236   -0.62657976  3.6200202 ], shape=(3,), dtype=float32)
txbeta:0 None
std_delta:0 tf.Tensor([0.3128128], shape=(1,), dtype=float32)
std_error:0 None
std_eta:0 tf.Tensor([7.756607], shape=(1,), dtype=float32)
std_nu:0 None

**Comment by xiaolong1979** [comment]
Seem to solve the problem by putting the trainable variables inside the model function.
